aws/resource_aws_appsync_graphql_api.go

```diff
@@ -0,0 +1,215 @@
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+func resourceAwsAppsyncGraphqlApi() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsAppsyncGraphqlApiCreate,
+		Read:   resourceAwsAppsyncGraphqlApiRead,
+		Update: resourceAwsAppsyncGraphqlApiUpdate,
+		Delete: resourceAwsAppsyncGraphqlApiDelete,
+
+		Schema: map[string]*schema.Schema{
+			"authentication_type": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := strings.ToUpper(v.(string))
+					validTypes := []string{"API_KEY", "AWS_IAM", "AMAZON_COGNITO_USER_POOLS"}
+					for _, str := range validTypes {
+						if value == str {
+							return
+						}
+					}
+					errors = append(errors, fmt.Errorf("expected %s to be one of %v, got %s", k, validTypes, value))
+					return
+				},
+				StateFunc: func(v interface{}) string {
+					return strings.ToUpper(v.(string))
+				},
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+					value := v.(string)
+					if !regexp.MustCompile(`[_A-Za-z][_0-9A-Za-z]*`).MatchString(value) {
+						errors = append(errors, fmt.Errorf("%q must match [_A-Za-z][_0-9A-Za-z]*", k))
+					}
+					return
+				},
+			},
+			"user_pool_config": {
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"app_id_client_regex": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"aws_region": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"default_action": {
+							Type:     schema.TypeString,
+							Required: true,
+							ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
+								value := strings.ToUpper(v.(string))
+								validTypes := []string{"ALLOW", "DENY"}
+								for _, str := range validTypes {
+									if value == str {
+										return
+									}
+								}
+								errors = append(errors, fmt.Errorf("expected %s to be one of %v, got %s", k, validTypes, value))
+								return
+							},
+							StateFunc: func(v interface{}) string {
+								return strings.ToUpper(v.(string))
+							},
+						},
+						"user_pool_id": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+					},
+				},
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceAwsAppsyncGraphqlApiCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).appsyncconn
+
+	input := &appsync.CreateGraphqlApiInput{
+		AuthenticationType: aws.String(d.Get("authentication_type").(string)),
+		Name:               aws.String(d.Get("name").(string)),
+		UserPoolConfig:     expandAppsyncGraphqlApiUserPoolConfig(d.Get("user_pool_config").([]interface{})),
```

Comment:
I am admittedly surprised this isn't causing a crash in the attached test where `user_pool_config` is not present. It's no big deal as it's not crashing, but it may be _safer_ to always wrap optional fields in `d.GetOk()`.